### ISSUE TITLE
[Feature request] Adding scaffold parameter estimator heads

### ISSUE THREAD

**Issue by eisenjulian** [tensorflow/tensorflow#16990]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: 1.4
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
When creating custom estimators, using the Heads defined [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/canned/head.py) reduces an important part of the boilerplate as well as guarantee uniform evaluation against canned estimator instances. However, one key parameter is missing in in `create_estimator_spec` function, which is the `scaffold` that one can pass into an `EstimatorSpec`. One example where that is needed is if you want to initialize a large tensor with a numpy array, for example for loading an embedding file from word2vec. This [StackOverflow question](https://stackoverflow.com/questions/44680769/loading-pre-trained-word2vec-to-initialise-embedding-lookup-in-the-estimator-mod) also describes the issue.

### Source code / logs
This is my normal code, currently impossible to use with heads. Actually, it might not be impossible, but this seems to me like the cleaner solution. I have the [tiny PR](https://github.com/eisenjulian/tensorflow/commit/9e6c004f07d2ba5c17d58007092f513b54077198) ready if you think this is a valuable change. Thanks a lot for your time!

```python
def model_fn(mode, features, labels, hparams):
  embed_ph = tf.placeholder(
      shape=[hparams.vocab_size, hparams.embedding_size], 
      dtype=tf.float32)
  embeddings = tf.Variable(embed_ph)
  # Define your model
  return tf.estimator.EstimatorSpec(
      ..., # normal EstimatorSpec args
      scaffold=tf.train.Scaffold(init_feed_dict={embed_ph: my_embedding_numpy_array})
  )
```

**Comment by cy89** [comment]
@martinwicke @ispirmustafa should we mark "contributions welcome"?

**Comment by martinwicke** [comment]
Mustafa, this looks correct to me, but your call.

**Comment by ispirmustafa** [comment]
playing with scaffold is an power user category. for power users I think following is a better approach:
est_spec = head.create_estimator_spec(...)
return est_spec._replace(scaffold=...)
I'm in favor of not adding scaffold into head interface.

**Comment by cy89** [comment]
@eisenjulian will @ispirmustafa 's suggestion get you, as a power user, going? If so, it's my inclination to close the discussion. But please advise (or close yourself if you're satisfied).

**Comment by eisenjulian** [comment]
Thanks for the suggestion: indeed I found about that approach by digging through the code and got to a working solution 👍. I can go ahead and close this.

Do you think there is value in updating the docs on [creating custom estimators](https://www.tensorflow.org/get_started/custom_estimators) regarding this idea? I mean both using `scaffold` for initializing large tensors and the `head` module. I am happy to help with that if you do. Loading pre-trained embeddings seems like a common enough use-case to account for and explain IMO.

For some context I'm a Google Developer Expert, working in collaboration with Google Dev Advocates to write a blog post on using estimators for NLP tasks.

**Comment by ispirmustafa** [comment]
for pre-trained embedding, I think warm_start_from is a better approach. Could you please check it out. it's a new argument we added to Estimator interface.

**Comment by eisenjulian** [comment]
Interesting, didn't know about this upcoming change, does this allow initializing with a `numpy` array? Or should we create dummy checkpoints from embeddings `.bin` or `.txt` files in advanced?

**Comment by ispirmustafa** [comment]
it works on checkpoint files. if you have numpy then scaffold is a better approach.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by eisenjulian** [comment]
I think this can be closed. For the record, I ended up finding even a better way of loading pre-trained embeddings this by using `tf.contrib.layers.embed_sequence` and passing in a custom initializer.

**Comment by AgoloCuongHoang** [comment]
Hi,

I just wanted to thanks @eisenjulian on his suggestion on load pre-trained embeddings using a custom initializer. You saved me a lot of time!
I didn't expect doing this is that hard with tf estimator.
